src/acquisition/covidcast/database.py

```diff
@@ -158,6 +163,179 @@ def update_direction(
+    '''
+
+    # A query that selects all rows from `covidcast` that have latest issue-date
+    # for any (time-series-key, time_value) with time_type='day'.
+    latest_issues_sql = f'''
+    SELECT 
+      `id`,
+      `source`,
+      `signal`,
+      `time_type`,
+      `geo_type`,
+      `geo_value`,
+      `time_value`,
+      `timestamp1`,
+      `value`,
+      `timestamp2`,
+      `direction`
+    FROM
+    (
+      SELECT
+        `source`,
+        `signal`,
+        `time_type`,
+        `geo_type`,
+        `geo_value`,
+        `time_value`,
+        MAX(`issue`) AS `issue`
+      FROM `covidcast`
+      WHERE
+        `time_type` = 'day'
+      GROUP BY
+        `source`,
+        `signal`,
+        `time_type`,
+        `geo_type`,
+        `geo_value`,
+        `time_value`
+    ) b
+    LEFT JOIN `covidcast` a
+    USING (`source`, `signal`, `time_type`, `geo_type`, `geo_value`, `time_value`, `issue`)
+    '''
+
+    cte_definition = f'''
+    WITH `latest_issues` AS
+    (
+      {latest_issues_sql}
+    )
+    '''
+
+    # A query that selects the keys of time-series with potentially stale directions.
+    stale_ts_key_sql = f'''
+      SELECT
+        `source`,
+        `signal`,
+        `time_type`,
+        `geo_type`,
+        `geo_value`
+      FROM
+        `latest_issues` AS t1
+      GROUP BY
+        `source`,
+        `signal`,
+        `time_type`,
+        `geo_type`,
+        `geo_value`
+      HAVING
+        MAX(`timestamp1`) > MIN(`timestamp2`)
+    '''
+
+    # A query that selects rows of the time-series selected by stale_ts_key_sql query.
+    stale_ts_records_with_latest_issues_sql = f'''
+    {cte_definition}
+    SELECT
+      `id`,
+      `source`,
+      `signal`,
+      `time_type`,
+      `geo_type`,
+      `geo_value`,
+      `time_value`,
+      `timestamp1`,
+      `value`,
+      `timestamp2`,
+      `direction`
+    FROM ({stale_ts_key_sql}) AS t2
+    LEFT JOIN `latest_issues` AS t3
+    USING (`source`, `signal`, `time_type`, `geo_type`, `geo_value`)
+    '''
+
+    if temporary_table is None:
+      self._cursor.execute(stale_ts_records_with_latest_issues_sql)
+      return list(self._cursor)
+
+    self._cursor.execute(create_tmp_table_sql)
+    final_sql = f'''
+      INSERT INTO `{temporary_table}`
+      SELECT *
+      FROM ({stale_ts_records_with_latest_issues_sql}) AS c;
+      '''
+    self._cursor.execute(final_sql)
+    self._cursor.execute(f"SELECT * FROM `{temporary_table}`;")
+    return list(self._cursor)
+
+  def batched_update_direction(self, new_direction_value, id_list, batch_size=1024):
+    """Updates the `direction` column of rows with ids in `id_list` to `new_direction_value`,
+    the update query is executed in batches of `bath_size` rows at a time.
+
+    `new_direction_value`: the new value to be stored in `direction` column.
+    `id_list`: a list of ids of rows that will change to `new_direction_value`.
+    `batch_size`: batch size of the update query.
+    """
```

Comment:
Can you add a brief explanation here on why batch_size is necessary, and guidance on how to decide whether to use the default or specify a higher or lower value?

Reply:
The reason why I added `batch_size` is that there is a maximum limit on the length of an SQL statement, and so in the case where the list size is 4M this limit will be certainly exceeded. As for choosing the `batch_size` there is a tradeoff, as with larger `batch_size` the update query will take longer to execute but there will be fewer queries. The only way I can think of to choose `batch_size` is experimentally. I see your point that it might be more appropriate for it to be hardcoded than to be exposed as a parameter.